xbmc/addons/AddonRepos.cpp

```diff
@@ -0,0 +1,214 @@
+  };
+
+  return addon->Origin() == ORIGIN_SYSTEM ||
+         std::any_of(officialRepoInfos.begin(), officialRepoInfos.end(), comparator);
+}
+
+void CAddonRepos::LoadAddonsFromDatabase(const CAddonDatabase& database)
+{
+  m_allAddons.clear();
+  database.GetRepositoryContent(m_allAddons);
+
+  m_addonsByRepoMap.clear();
+  for (const auto& addon : m_allAddons)
+  {
+    if (m_addonMgr.IsCompatible(*addon))
+    {
+      m_addonsByRepoMap[addon->Origin()].insert({addon->ID(), addon});
+    }
+  }
+
+  for (const auto& map : m_addonsByRepoMap)
+    CLog::Log(LOGDEBUG, "ADDONS: repo: {} - {} addon(s) loaded", map.first, map.second.size());
+
+  SetupLatestVersionMaps();
+}
+
+void CAddonRepos::SetupLatestVersionMaps()
+{
+  m_latestOfficialVersions.clear();
+  m_latestPrivateVersions.clear();
+  m_latestVersionsByRepo.clear();
+
+  for (const auto& repo : m_addonsByRepoMap)
+  {
+    const auto& addonsPerRepo = repo.second;
+
+    for (const auto& addonMapEntry : addonsPerRepo)
+    {
+      const auto& addonToAdd = addonMapEntry.second;
+
+      if (IsFromOfficialRepo(addonToAdd, true))
+      {
+        AddAddonIfLatest(addonToAdd, m_latestOfficialVersions);
+      }
+      else
+      {
+        AddAddonIfLatest(addonToAdd, m_latestPrivateVersions);
+      }
+
+      // add to latestVersionsByRepo
+      AddAddonIfLatest(repo.first, addonToAdd, m_latestVersionsByRepo);
+    }
+  }
+}
+
+void CAddonRepos::AddAddonIfLatest(const AddonPtr& addonToAdd,
+                                   std::map<std::string, AddonPtr>& map) const
+{
+  const auto& latestKnown = map.find(addonToAdd->ID());
+  if (latestKnown == map.end() || addonToAdd->Version() > latestKnown->second->Version())
+    map[addonToAdd->ID()] = addonToAdd;
+}
+
+void CAddonRepos::AddAddonIfLatest(
+    const std::string& repoId,
+    const AddonPtr& addonToAdd,
+    std::map<std::string, std::map<std::string, AddonPtr>>& map) const
+{
+  const auto& latestVersionByRepo = map.find(repoId);
+
+  if (latestVersionByRepo == map.end()) // repo not found
+  {
+    map[repoId].insert({addonToAdd->ID(), addonToAdd});
+  }
+  else
+  {
+    const auto& latestVersionEntryByRepo = latestVersionByRepo->second;
+    const auto& latestKnown = latestVersionEntryByRepo.find(addonToAdd->ID());
+
+    if (latestKnown == latestVersionEntryByRepo.end() ||
+        addonToAdd->Version() > latestKnown->second->Version())
+      map[repoId][addonToAdd->ID()] = addonToAdd;
+  }
+}
+
+void CAddonRepos::BuildUpdateList(const std::vector<AddonPtr>& installed,
+                                  std::vector<AddonPtr>& updates) const
+{
+  AddonPtr update;
+
+  for (const auto& addon : installed)
+  {
+    CLog::Log(LOGDEBUG, "ADDONS: update check: addonID = {} / Origin = {}", addon->ID(),
+              addon->Origin());
+
+    if (DoAddonUpdateCheck(addon, update))
+      updates.emplace_back(update);
+  }
+}
+
+bool CAddonRepos::DoAddonUpdateCheck(const AddonPtr& addon, AddonPtr& update) const
```

Comment:
first i wanted to return the pointer but then switched to bool because of this:
https://github.com/xbmc/xbmc/blob/6372cf6f4ff8d8444a40674d3730c35392bdc429/xbmc/addons/AddonRepos.cpp#L180-L181